blaze/expr/core.py

```diff
@@ -404,6 +404,7 @@ def subterms(x):
     yield x
 
 
+@memoize_mutable
 def subs(o, d):
```

Comment:
I am pretty sure there is a dispatch for `Mono` to swap out types in a datashape, mainly type or dim vars. Is subs that expensive that we want to add this kind of memoization? The serialization/deserialization seem much more expensive than a tree walk.